lib/captain_hook.rb

```diff
@@ -44,17 +44,19 @@ def hook(
     # Hooks logic part
     ####
     def get_hooks(kind)
-      # Only get hooks from the most specific class that defines them
-      return hooks[kind].values if hooks[kind].any?
+      hook_list = Hash.new { |h, k| h[k] = [] }
 
-      # If no hooks defined in this class, look up the inheritance chain
-      ancestors[1..].each do |ancestor|
-        next unless ancestor.respond_to?(:hooks)
-        return ancestor.hooks[kind].values if ancestor.hooks[kind].any?
+      # Collect hooks from the current class and ancestor classes
+      [self, *ancestors[1..]].each do |klass|
+        next unless klass.respond_to?(:hooks)
+
+        (klass.hooks[kind] || {}).each_value do |hook|
+          hook_list[hook.hook.class] << hook
+        end
       end
 
-      # If no hooks found anywhere in the chain, return empty array
-      []
+      # Return an array containing the first element of each key in hook_list
+      hook_list.values.map(&:first)
```

Comment:
why `&:first`? how this `hook_list.values` looks like?

Reply:
On each iteration of the loop, we stored all the occurrences of hooks from a class and its ancestors in the `hook_list`. Since the array is sorted, selecting the first element means that we retrieve the hook definition closer to the child class.

Considering the following class definitions:: 

```
class Child
   hook :before, hook: Hook1.new, ...
end

class Parent < Child
   hook :before, hook: Hook1.new, ...
   hook :before, hook: Hook2.new, ...
end
```

Calling `get_hooks(:before)` will set the value of `hook_list` to this:
```
{
  "Hook1": [Hook1 (from Child), Hook1 (from Parent),
  "Hook2": [Hook2 (from Parent)]
}
```

and `hook_list.values` will looks like:

```
[
[Hook1 (from Child), Hook1 (from Parent)],
[Hook2 (from Parent)]
]
```

So mapping the values using `first` will produce an array like:

```
[Hook1 (from Child), Hook2 (from Parent)]
```